compositor_pipeline/src/queue/internal_queue.rs

```diff
@@ -0,0 +1,136 @@
+use anyhow::anyhow;
+use compositor_common::frame::{Framerate, FramesBatch, InputID, Pts};
+use compositor_common::Frame;
+use std::collections::hash_map::Entry::Vacant;
+use std::time::Duration;
+use std::{collections::HashMap, sync::Arc};
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum QueueError {
+    #[error("the input id `{0}` is unknown")]
+    UnknownInputID(InputID),
+}
+
+pub struct InternalQueue {
+    // frames are PTS ordered. PTS include timestamps offsets
+    inputs_queues: HashMap<InputID, Vec<Arc<Frame>>>,
+    timestamp_offsets: HashMap<InputID, Pts>,
+    output_framerate: Framerate,
+    pub send_batches_counter: u32,
+}
+
+impl InternalQueue {
+    pub fn new(output_framerate: Framerate) -> Self {
+        InternalQueue {
+            inputs_queues: HashMap::new(),
+            timestamp_offsets: HashMap::new(),
+            output_framerate,
+            send_batches_counter: 0,
+        }
+    }
+
+    pub fn add_input(&mut self, input_id: InputID) {
+        self.inputs_queues.insert(input_id, Vec::new());
+    }
+
+    pub fn remove_input(&mut self, input_id: InputID) {
+        self.inputs_queues.remove(&input_id);
+        self.timestamp_offsets.remove(&input_id);
+    }
+
+    pub fn enqueue_frame(&mut self, input_id: InputID, mut frame: Frame) -> Result<(), QueueError> {
+        match self.inputs_queues.get_mut(&input_id) {
+            Some(input_queue) => {
+                if let Vacant(e) = self.timestamp_offsets.entry(input_id) {
+                    e.insert(frame.pts);
+                }
+
+                frame.pts += *self
+                    .timestamp_offsets
+                    .get(&input_id)
+                    .ok_or_else(|| anyhow!("Timestamp offset unregistered for pad {input_id}"))
```

Comment:
no strong opinion, but I think we wanted to use anyhow only in top-level crates, and stick to precise errors in lib packages